src/chunker/disprove_execution.rs

```diff
@@ -0,0 +1,301 @@
+use std::rc::Rc;
+
+use ark_groth16::{Proof, VerifyingKey};
+
+use super::{
+    assigner::BCAssigner, chunk_groth16_verifier::groth16_verify_to_segments, common::RawWitness,
+    elements::dummy_element,
+};
+
+#[derive(Clone)]
+pub struct RawProof {
+    pub proof: Proof<ark_bn254::Bn254>,
+    pub public: Vec<<ark_bn254::Bn254 as ark_ec::pairing::Pairing>::ScalarField>,
+    pub vk: VerifyingKey<ark_bn254::Bn254>,
+}
+
+pub fn disprove_exec<A: BCAssigner>(
+    assigner: &A,
+    assert_witness: Vec<Vec<RawWitness>>,
+    wrong_proof: RawProof,
+) -> Option<(usize, RawWitness)> {
+    // 0. TODO: if 'wrong_proof' is correct, return none
+
+    // 1. derive assigner from wrong proof
+    let mut wrong_proof_assigner = A::default();
+    let mut segments = groth16_verify_to_segments(
+        &mut wrong_proof_assigner,
+        &wrong_proof.public,
+        &wrong_proof.proof,
+        &wrong_proof.vk,
+    );
+    let segment_length = segments.len();
+
+    // 2. recover assigner from witness
+    let hash_map = assigner.recover_from_witness(assert_witness);
+
+    // 3. find which chunk is unconsistent
+    for (idx, segment) in segments.iter_mut().enumerate() {
+        let mut is_param_equal = true;
+        for param in segment.parameter_list.iter() {
+            if param.to_hash().unwrap() != hash_map.get(param.id()).unwrap().clone() {
+                is_param_equal = false;
+            }
+        }
+        let mut is_result_equal = true;
+        for result in segment.result_list.iter_mut() {
+            if result.to_hash().unwrap() != hash_map.get(result.id()).unwrap().clone() {
+                is_result_equal = false;
+                // replace the result to hash_map
+                *result = Rc::new(Box::new(dummy_element(
+                    &result.id(),
+                    hash_map.get(result.id()).unwrap().clone(),
+                )));
```

Comment:
This function is used when the operator has released all intermediate state hashes, and a segment will first check bit commitment and then run the chunk. So for the inconsistent `result`, it needs to replace all inconsistent results to make sure that the bit commitment check can pass.

The incorrect `param` is useless, because we don't know the behavior of segment with incorrect `param`s. But, we can always find the segment containing consistent `param` and inconsistent `result`.

Reply:
I see. So the verifier is supposed to create its own `assigner` by executing the segments without the Winternitz part and retrieve the hashes that way. Then they can check for errors in the bit commited hashes from the operator. Do I understand that correctly?

Reply:
Yes, you're right. A different type of assigner is needed as a challenger. Running the chunk one by one is another obvious solution, but I think this way is more efficient.